src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -278,94 +317,57 @@ private void FreeKey()
             }
         }
 
-        private static int GetKeySize(SafeEcKeyHandle ecKeyHandle)
-        {
-            int nid = Interop.Crypto.EcKeyGetCurveName(ecKeyHandle);
-            int keySize = 0;
-
-            for (int i = 0; i < s_supportedAlgorithms.Length; i++)
-            {
-                if (s_supportedAlgorithms[i].Nid == nid)
-                {
-                    keySize = s_supportedAlgorithms[i].KeySize;
-                    break;
-                }
-            }
-
-            if (keySize == 0)
-            {
-                string curveNameOid = Interop.Crypto.GetOidValue(Interop.Crypto.ObjNid2Obj(nid));
-                throw new NotSupportedException(SR.Format(SR.Cryptography_UnsupportedEcKeyAlgorithm, curveNameOid));
-            }
-
-            return keySize;
-        }
-
         private SafeEcKeyHandle GenerateKey()
         {
-            int keySize = KeySize;
-            for (int i = 0; i < s_supportedAlgorithms.Length; i++)
+            if (this.Curve != null)
             {
-                if (keySize == s_supportedAlgorithms[i].KeySize)
-                {
-                    int nid = s_supportedAlgorithms[i].Nid;
-                    SafeEcKeyHandle key = Interop.Crypto.EcKeyCreateByCurveName(nid);
-                    if (key == null || key.IsInvalid)
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                int nid = Interop.Crypto.OidToNid(Curve);
 
-                    if (!Interop.Crypto.EcKeyGenerateKey(key))
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                SafeEcKeyHandle key = Interop.Crypto.EcKeyCreateByCurveName(nid);
 
-                    return key;
-                }
-            }
+                if (key == null || key.IsInvalid)
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
 
-            // The KeySize property should have prevented a bad KeySize from being set.
-            Debug.Fail("GenerateKey: Unexpected KeySize: " + keySize);
-            throw new InvalidOperationException();  // This is to keep the compiler happy - we don't expect to hit this.
-        }
+                if (!Interop.Crypto.EcKeyGenerateKey(key))
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
 
-        private Lazy<SafeEcKeyHandle> _key;
+                _tempKeySize = Interop.Crypto.EcKeyGetSize(key);
```

Comment:
Like we did in RSACng/RSAOpenSsl; it'd be good to just make whatever needs to happen here be a method, so it isn't more than one line being copied over and over.